### PR TITLE
separate coulomb contribution calculation from nse solver

### PR DESCRIPTION
**Pull request by zhichen3**
Separate out calculation of coulomb contribution from nse solver to make things faster.

Even though coulomb contribution depends on electron fraction, `y_e`, but the `y_e` we use to evaluate the coulomb contribution does not change during iteration. We just use the input `y_e`.

after #1507

### CODE REVIEW

**Comment by zhichen3** [comment]
new profile using this pr, which can be compared with the ones in #1506. I'm using strang for testing btw

```
time   seconds   seconds    calls   s/call   s/call  name    
 14.82     47.35    47.35 388320371     0.00     0.00  std::_Function_handler<amrex::GpuTuple<amrex::ValLocPair<double, amrex::IntVect> > (), amrex::ReduceData<amrex::ValLocPair<double, amrex::IntVect> >::ReduceData<amrex::ReduceOpMin>(amrex::ReduceOps<amrex::ReduceOpMin>&)::{lambda()#1}>::_M_manager(std::_Any_data&, std::_Any_data const&, std::_Manager_operation)
 13.37     90.07    42.72 27737138     0.00     0.00  void evaluate_rates<0, rate_t>(burn_t const&, rate_t&)
  8.44    117.03    26.96 887588416     0.00     0.00  get_partition_function(int, tf_t const&, double&, double&)
  8.02    142.66    25.63   227724     0.00     0.00  void actual_integrator<burn_t>(burn_t&, double, bool)
  6.24    162.60    19.94 153867861     0.00     0.00  std::_Function_handler<amrex::GpuTuple<amrex::ValLocPair<double, amrex::IntVect> > (), amrex::ReduceData<amrex::ValLocPair<double, amrex::IntVect> >::ReduceData<amrex::ReduceOpMin>(amrex::ReduceOps<amrex::ReduceOpMin>&)::{lambda()#1}>::_M_invoke(std::_Any_data const&)
  5.84    181.24    18.64  8930532     0.00     0.00  burn_t get_actual_nse_state<burn_t>(burn_t&, double, bool)
  5.28    198.11    16.87 90989518     0.00     0.00  void apply_electrons<burn_t>(burn_t&)
  4.53    212.59    14.48 44008987     0.00     0.00  void fcn<burn_t>(amrex::Array1D<double, 1, 2>&, amrex::Array1D<double, 1, 2>&, nse_solver_data<burn_t> const&, int&) [clone .constprop.1]
  4.16    225.87    13.28 30856531     0.00     0.00  void actual_eos<eos_input_t, burn_t>(eos_input_t, burn_t&)
  4.04    238.77    12.90  8949977     0.00     0.00  void hybrj<2, nse_solver_data<burn_t> >(hybrj_t<2>&, nse_solver_data<burn_t> const&)
  4.00    251.55    12.78  5973880     0.00     0.00  nse_grouping(amrex::Array1D<int, 1, 18>&, burn_t const&, amrex::Array1D<double, 1, 18> const&, amrex::Array1D<double, 1, 85> const&, double)
  2.70    260.18     8.63 27737138     0.00     0.00  void fill_reaclib_rates<0, rate_t>(tf_t const&, rate_t
```

File: util/microphysics_sort.H [hunk 1]
```diff
@@ -1,23 +1,99 @@
...
+    else {
+        return a > b;
+    }
+}
+
+
+template <typename T>
 AMREX_GPU_HOST_DEVICE AMREX_INLINE
-void sort_Array1D(amrex::Array1D<int, l, m>& array_1d){
+void swap(T& a, T& b) {
```
Comment (yut23) [inline util/microphysics_sort.H]: Can this be replaced with `std::swap()`?

File: util/microphysics_sort.H [hunk 2]
```diff
@@ -1,23 +1,99 @@
...
+}
+
+template <typename T, int l, int m>
+AMREX_GPU_HOST_DEVICE AMREX_INLINE
+void quickSort_Array1D(amrex::Array1D<T, l, m>& array_1d,
+                       const int low, const int high,
+                       const bool ascending=true) {
+    // quickSort implementation
+    // This version uses the array itself as the metric for sorting
+    if (low < high) {
```
Comment (yut23) [inline util/microphysics_sort.H]: This should check `compare(low, high, ascending)` instead

File: util/microphysics_sort.H (hunk 1 above)
Comment (zhichen3) [reply util/microphysics_sort.H]: I have the impression that standard library in general don't work well with cuda/gpu, but I'm not sure.

File: util/microphysics_sort.H (hunk 2 above)
Comment (zhichen3) [reply util/microphysics_sort.H]: I think the current version on development works fine for both `ascending=true` and `ascending=false`. I just tested it using a random array1d and the results seem to be right.

Comment (yut23) [reply util/microphysics_sort.H]: You're right. I was reading it wrong. I'm also not sure why I wasn't reviewing the latest version of the PR.

**Comment by zingale** [comment]
are we happy with this now?

**Review by zingale — APPROVED** [review]

**Pull request merged**